store/mongo/datastore_mongo.go

```diff
@@ -368,22 +375,47 @@ func (db *DataStoreMongo) upsertAttributes(
 	case 1:
 		var res *mongo.UpdateResult
 		if withRevision {
-			filter = bson.M{
-				"_id":         devices[0].Id,
-				DbDevRevision: bson.M{"$lt": devices[0].Revision},
+			if etag != nil && *etag != "" {
+				filter = bson.M{
+					"_id":         devices[0].Id,
+					DbDevRevision: bson.M{"$lt": devices[0].Revision},
+					etagField:     bson.M{"$exists": true, "$eq": *etag},
+				}
+				if err := checkMatchedCount(ctx, c, filter); err != nil {
+					return &model.UpdateResult{MatchedCount: 0}, nil
+				}
+
+			} else {
+				filter = bson.M{
+					"_id":         devices[0].Id,
+					DbDevRevision: bson.M{"$lt": devices[0].Revision},
+				}
 			}
 			update[DbDevRevision] = devices[0].Revision
-			update = bson.M{
-				"$set":         update,
-				"$setOnInsert": oninsert,
-			}
 		} else {
-			filter = map[string]interface{}{"_id": devices[0].Id}
-			update = bson.M{
-				"$set":         update,
-				"$setOnInsert": oninsert,
+			if etag != nil && *etag != "" {
+				filter = bson.M{
+					"_id":     devices[0].Id,
+					etagField: bson.M{"$exists": true, "$eq": *etag},
+				}
+				// check if we are going to update existing document with the
+				// filter, otherwise with SetUpsert(true) the query ends up
+				// with 'duplicate key' error
+				if err := checkMatchedCount(ctx, c, filter); err != nil {
+					return &model.UpdateResult{MatchedCount: 0}, nil
+				}
+			} else {
+				filter = map[string]interface{}{"_id": devices[0].Id}
 			}
 		}
+		if scope == model.AttrScopeTags {
+			update[etagField] = fmt.Sprintf("%s", uuid.New())
```

Comment:
uuid objects have a method to return a string (is it .String(), maybe?)

Reply:
👍